### DISCUSSION TITLE
Able to get a vector without 'vocab', but no vector with 'vocab'

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#7400] [Help: Coding & Implementations] [answered]
Hi! I just started learning spaCy and I am not understanding one particular vector behavior.
My pipeline consists of NER and tok2vec components. I pretrained on top of en_core_web_lg, and included pretrained vectors as well as vocab_data.jsonl:

config.cfg section:
```
[initialize]
vectors = "en_core_web_lg"
init_tok2vec = "./pretrained_vectors/model999.bin"
vocab_data = "./vocab_data.jsonl"
lookups = null
before_init = null
after_init = null
``` 
After I trained my pipeline on my annotated data and checked the vector for the word that is both in the annotated labels as well as in the vocab_data.jsonl, it says it has no vector, and it's OOV. (and here I thought that adding a word to vocab_data will add it to the vocab)
However, when I delete the 'vocab' folder from my pipeline, then it says that there exists a vector for that word if I use this way:

```
nlp('myword')[0].has_vector
nlp('myword')[0].vector
```

My questions are:
1. Why `nlp('myword')[0].vector`  returns no vector when 'vocab' is there and returns a vector when 'vocab' is deleted
2. Why are the words from vocab_data.jsonl or from my trained annotations not reflected in the vectors

Thank you!

**Comment by adrianeboyd** [comment, marked as answer]
If you delete the `vocab` folder the model shouldn't be able to be loaded at all, so I'm not sure what's going on there.

The definition of `is_oov` changed from "doesn't have prob/cluster features in lexeme data" (`vocab_data`) to "doesn't have a vector" between v2.2 and v2.3 because most of the lexeme data is no longer saved with the model. All the derivable features like `suffix` are calculated on the fly rather than being saved/loaded.

If you have vectors (`nlp.vocab.vectors.size > 0`), then `nlp("word")[0].vector` should be the same as `nlp.vocab["word"].vector`. The vectors are not part of `vocab_data`. With this config, in terms of the vectors you should see the exact same behavior for a word in your new pipeline as for a word in `en_core_web_lg`.

`vocab_data` is mainly for backwards compatibility for people who have some of this data in the old format from previous versions of spacy (it was more important in spacy v1.x) and still want to set `norm` / `prob` / `cluster` this way rather than through lookups tables. In most cases, you probably don't need it. You definitely don't need it to train any of the built-in components like NER, which haven't used `cluster` or `prob` as features since spacy v2.0+.

Ah, I think I see part of the confusion. The `is_oov` from `vocab_data` isn't actually used because in v2.3+ it only looks for the vector, not this lexeme flag. Again, this table is more for backwards compatibility than anything. Try training without it?

**Reply by ghost** [reply]
@adrianeboyd Thank you for a quick reply! Apologies for not mentioning that I used spaCy 3.0. Removing 'vocab' folder is not breaking anything for me, just behaves as if I trained on top of blank:xx.

Since adding vocab_data doesn't affect the vectors, and the training/pretraining doesn't affect the vectors, how do I add new vectors to the existing vocab from en_core_web_lg? I know I could use Gensim to generate my own vectors, but can I add them directly with vectors.add or merge them with en_core_web_lg somehow?  I want to run similarity function on data that has a mix of English words and cli commands, so I need both en_core_web_lg vectors plus my custom ones, and I am looking for the right approach.

Thanks again!

**Reply by adrianeboyd** [reply]
You mean removing `vocab_data` and have `vocab_data = null`? Yes, that's fine. (You can't remove the `vocab/` folder from a saved model directory.)

It may not make sense to combine your custom vectors with the glove vectors in en_core_web_lg. Because the vector spaces aren't aligned, it might be better to train new vectors just on your data and create a new vectors model with `spacy init vectors`.

If you do want to extend `en_core_web_lg`, add the vectors with `vectors.add` and save the model in a new directory like `en_core_web_lg_custom`. Then use the path to this directory as `paths.vectors` instead of the model name. It is easy to add a new token with the exact same vector as an existing token. This doesn't even require more data in the model because it can point to the same row in the vectors table:

```python
nlp.vocab.vectors.add("dog", row=nlp.vocab.vectors.key2row[nlp.vocab.strings["cat"]])
```

This points "dog" to the existing row for "cat".

**Reply by ghost** [reply]
@adrianeboyd  Maybe I did something wrong, but model without 'vocab' folder works just fine for me, except it doesn't have vectors in nlp.vocab, which makes sense.

Thanks for clarification, looks like I will have to train new vectors on my mixed english/code data, not sure if it will be any good, considering code/commands don't have sentence structure and context that english sentences do, but will give it a try. 

Thank you!
